include/llvm-dialects/Dialect/OpMap.h

```diff
@@ -0,0 +1,356 @@
+
+    return {};
+  }
+
+  // -------------------------------------------------------------
+  // Inserts a given Key/Value pair and return a reference or
+  // directly return a reference if the key already exists.
+  // -------------------------------------------------------------
+
+  ValueT &insert_or_get(const OpDescription &desc, const ValueT &val) {
+    if (desc.isCoreOp() || desc.isIntrinsic()) {
+      assert(desc.getOpcodes().size() == 1 &&
+             "Can only insert a single op at a time.");
+
+      const unsigned op = desc.getOpcodes().front();
+      if (desc.isCoreOp()) {
+        m_coreOpcodes[op] = val;
+        return m_coreOpcodes[op];
+      }
+
+      m_intrinsics[op] = val;
+      return m_intrinsics[op];
+    }
+
+    for (DialectOpKV &dialectOpKV : m_dialectOps) {
+      if (dialectOpKV == desc) {
+        dialectOpKV.Value = val;
+        return dialectOpKV.Value;
+      }
+    }
+
+    m_dialectOps.push_back({getDialectMapKey(desc), val});
+    return m_dialectOps.back().Value;
+  }
+
+  template <typename OpT> ValueT &insert_or_get(const ValueT &val) {
+    const OpDescription desc = OpDescription::get<OpT>();
+    return insert_or_get(desc, val);
+  }
+
+  ValueT &insert_or_get(const OpDescription &desc, ValueT &&val) {
+    if (desc.isCoreOp() || desc.isIntrinsic()) {
+      assert(desc.getOpcodes().size() == 1 &&
+             "Can only insert a single op at a time.");
+
+      const unsigned op = desc.getOpcodes().front();
+      if (desc.isCoreOp()) {
+        m_coreOpcodes[op] = std::move(val);
+        return m_coreOpcodes[op];
+      }
+
+      m_intrinsics[op] = std::move(val);
+      return m_intrinsics[op];
+    }
+
+    for (DialectOpKV &dialectOpKV : m_dialectOps) {
+      if (dialectOpKV == desc) {
+        dialectOpKV.Value = std::move(val);
+        return dialectOpKV.Value;
+      }
+    }
+
+    m_dialectOps.push_back({getDialectMapKey(desc), std::move(val)});
+    return m_dialectOps.back().Value;
+  }
+
+  template <typename OpT> ValueT &insert_or_get(ValueT &&val) {
+    const OpDescription desc = OpDescription::get<OpT>();
+    return insert_or_get(desc, std::move(val));
+  }
+
+  // -------------------------------------------------------------
+  // Erase a given operation from the correct container.
+  // -------------------------------------------------------------
+
+  // Erase a given dialect operation.
+  template <typename OpT> bool erase() {
+    OpDescription desc = OpDescription::get<OpT>();
+    return erase(desc);
+  }
+
+  // Erase all the operations described by a given OpDescription.
+  bool erase(const OpDescription &desc) {
+    bool erased = false;
+
+    if (desc.isCoreOp()) {
+      for (const unsigned op : desc.getOpcodes())
+        erased |= m_coreOpcodes.erase(op);
+    } else if (desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes())
+        erased |= m_intrinsics.erase(op);
+    }
+
+    // Try to erase the dialect op at last.
```

Comment:
This should be under an `else` (or just always directly `return` from the cases above)